drivers/mtd/at25ee.c

```diff
@@ -0,0 +1,1033 @@
+{
+  at25ee_lock(priv->spi);
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), true);
+
+  SPI_SEND(priv->spi, enable ? AT25EE_CMD_WREN : AT25EE_CMD_WRDIS);
+
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), false);
+  at25ee_unlock(priv->spi);
+}
+
+/****************************************************************************
+ * Name: at25ee_writepage
+ *
+ * Description:
+ *   Write data to the EEPROM, NOT crossing page boundaries.
+ *
+ * Input Parameters:
+ *   priv        - a reference to the device structure
+ *   devaddr     - the address to start the write
+ *   data        - pointer to data buffer to write
+ *   len         - length of the data to write
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void at25ee_writepage(FAR struct at25ee_dev_s *priv, uint32_t devaddr,
+                             FAR const uint8_t *data, size_t len)
+{
+  at25ee_lock(priv->spi);
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), true);
+
+  at25ee_sendcmd(priv->spi, AT25EE_CMD_WRITE, priv->addrlen, devaddr);
+  SPI_SNDBLOCK(priv->spi, data, len);
+
+  SPI_SELECT(priv->spi, SPIDEV_EEPROM(0), false);
+  at25ee_unlock(priv->spi);
+}
+
+/****************************************************************************
+ * Name: at25ee_eraseall
+ *
+ * Description:
+ *   Erase all data in the device
+ *
+ * Input Parameters:
+ *   priv        - a reference to the device structure
+ *   devaddr     - the address to start the write
+ *   data        - pointer to data buffer to write
+ *   len         - length of the data to write
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static int at25ee_eraseall(FAR struct at25ee_dev_s *priv)
+{
+  uint8_t *buf;
+  int startblock = 0;
+
+  DEBUGASSERT(priv);
+
+  buf = kmm_malloc(priv->pgsize);
+  if (!buf)
+    {
+      ferr("ERROR: Failed to alloc memory for at25ee eraseall!\n");
+      return -ENOMEM;
+    }
+
+  for (startblock = 0; startblock < priv->npages; startblock++)
+    {
+      uint16_t offset = startblock * priv->pgsize;
+      at25ee_write(&priv->mtd, offset, priv->pgsize, buf);
+    }
+
+  kmm_free(buf);
+  return OK;
+}
+
+/****************************************************************************
+ * Name: at25ee_erase
+ *
+ * Description:
+ *   Erase a number of blocks of data.
+ *   NB - It is not necessary to erase data before writing with EEPROMs
+ *
+ * Input Parameters:
+ *   dev        - a reference to the device structure
+ *   startblock - start block of the erase
+ *   nblocks    - nblocks to erase
+ *
+ * Returned Value:
+ *   Success (OK) or fail (negated error code)
+ ****************************************************************************/
+
+static int at25ee_erase(FAR struct mtd_dev_s *dev,
+                       off_t startblock,
+                       size_t nblocks)
+{
+  return (int)nblocks;
```

Comment:
why not implement

Reply:
@xiaoxiang781216 Because you don't need to erase EEPROM before you write. As far as I can tell, we only need an eraseall function to wipe the memory before trying to format/mount a new FS - a bulk erase via ioctl. I don't think the erase is, actually, ever called. Not in the testing I've done.

The comment description block mentions this but I could include a specific comment in the function if liked?

Reply:
If at25 doesn't need erase before write, it's better to implement as a block device not mtd device.

Reply:
Ok, which file system do you want to use?

Reply:
So, the erase needs fill 0xff to the specified range.

Reply:
There definitely _was_ a memcpy there - must have lost it. That's the downside of squashing prior to commit - you lose the history (unless you take a copy of the branch as a sub branch...note to self!!).

I will implement the erase. Someone else can add it to the at24ee eeprom mtd driver as it doesn't have it either ;-)

Reply:
I added the blockerase when playing around with filesystems on this, and testing. If there was an error in my code and/or something went wrong with the mount or format, I had to do a bulk erase to get over it.

I didn't trace any calls to the page/block erase as I was, actually, mostly playing around with littlefs and fat. But obviously this driver should be generic so, as above, I will add the erase.

Reply:
done

Reply:
> Proper emulation of MTD on an eeprom requires erasing.
> 
> I suggest to implement it even if unnecessary, because it's not hard, and not doing so will confuse the hell out of people expecting erase to actually erase.
> 
> Let's say it's the specification of the MTD drivers and all implementations should do it.

Although this is now working right, I still don't 100% understand why we need it?

As you know, EEPROM does not need to be erased before you write. The comment made this clear, and is the same as the i2c at24ee mtd driver. It adds unnecessary time to the function and reduces the lifetime of the part.

BULK erase is a different matter as you would need/want do that before formatting with a new FS, for example, or reformatting an existing in case of FS corruption?

Reply:
@TimJTi is it done to be integrated into mainline? If so, please remove it from draft.

Reply:
I will, but was waiting on feedback regarding this erase issue which might not be until next year of course. I don't want it accepted and merged until there is consensus on this.

Reply:
I have made the block erase something that can be enabled via Kconfig if needed.